### PR TITLE
HBASE-22582 The Compaction writer may access the lastCell whose memor…

### PR DESCRIPTION
**Pull request by openinx**
…y has been released when appending fileInfo in the final

### CODE REVIEW

**Comment by openinx** [comment]
Will try to add few UT to address this bug.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 24 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -0 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 262 | master passed |
| +1 | compile | 57 | master passed |
| +1 | checkstyle | 69 | master passed |
| +1 | shadedjars | 267 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 202 | master passed |
| +1 | javadoc | 34 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 239 | the patch passed |
| +1 | compile | 53 | the patch passed |
| +1 | javac | 53 | the patch passed |
| +1 | checkstyle | 70 | hbase-server: The patch generated 0 new + 40 unchanged - 1 fixed = 40 total (was 41) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 263 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 726 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 215 | the patch passed |
| +1 | javadoc | 29 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 8288 | hbase-server in the patch passed. |
| +1 | asflicense | 29 | The patch does not generate ASF License warnings. |
| | | 11129 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/341 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 5c9935f9e0cb 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | master / 0198868531 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/1/testReport/ |
| Max. process+thread count | 4648 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by jatsakthi** [comment]
@openinx , I see some syntax/style fixes along with the "beforeShipped()" fix. I can confirm that there are other places as well in these 3 files were such changes could be welcomed. Do you think that we should include those as well here?

**Comment by openinx** [comment]
@jatsakthi  Yeah, I think we can fix the checksytle in this issue.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java [hunk 1]
```diff
@@ -331,6 +327,7 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
           "Interrupted while control throughput of compacting " + compactionName);
     } finally {
       throughputController.finish(compactionName);
+      ((ShipperListener) writer).beforeShipped();
```
Comment (anoopsjohn) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: Below in Compactor, call beforeShipped() before finishing the throughputController. Its ok for any order. Still we can maintain one order.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java [hunk 2]
```diff
@@ -311,13 +311,9 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
...
         if (LOG.isDebugEnabled()) {
           if ((now - lastMillis) >= COMPACTION_PROGRESS_LOG_INTERVAL) {
-            LOG.debug("Compaction progress: "
-                + compactionName
-                + " "
-                + progress
-                + String.format(", rate=%.2f kB/sec", (bytesWrittenProgressForLog / 1024.0)
-                    / ((now - lastMillis) / 1000.0)) + ", throughputController is "
-                + throughputController);
+            double rate = (bytesWrittenProgressForLog / 1024.0) / ((now - lastMillis) / 1000.0);
```
Comment (anoopsjohn) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: What abt the String formatting on the Double value which we were doing.  Missing that in log now.  Below one more place too.

Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: Emm... maybe I need to format the rate first before put it as a arg in LOG.debug, thanks .

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java (hunk 1 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: Yeah,  I will provide a UT to address this bug but am working on other things now... will update this patch later.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 213 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 312 | master passed |
| +1 | compile | 55 | master passed |
| +1 | checkstyle | 70 | master passed |
| +1 | shadedjars | 285 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 218 | master passed |
| +1 | javadoc | 32 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 259 | the patch passed |
| +1 | compile | 51 | the patch passed |
| +1 | javac | 51 | the patch passed |
| +1 | checkstyle | 68 | hbase-server: The patch generated 0 new + 66 unchanged - 1 fixed = 66 total (was 67) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 259 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 782 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 203 | the patch passed |
| +1 | javadoc | 30 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 15853 | hbase-server in the patch failed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 19050 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.TestAssignmentManagerMetrics |
|   | hadoop.hbase.master.TestSplitWALManager |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/341 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 516e633f3a5b 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | master / bdf9d56f2d |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/2/testReport/ |
| Max. process+thread count | 5305 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java [hunk 3]
```diff
@@ -311,13 +311,10 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
...
         if (LOG.isDebugEnabled()) {
           if ((now - lastMillis) >= COMPACTION_PROGRESS_LOG_INTERVAL) {
-            LOG.debug("Compaction progress: "
-                + compactionName
-                + " "
-                + progress
-                + String.format(", rate=%.2f kB/sec", (bytesWrittenProgressForLog / 1024.0)
-                    / ((now - lastMillis) / 1000.0)) + ", throughputController is "
-                + throughputController);
+            String rate = String.format("%.2f",
```
Comment (infraio) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: if (LOG.isDebugEnabled())? Then can skip to generate "rate" String in production environment.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java [hunk 4]
```diff
@@ -447,13 +448,10 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
...
         if (LOG.isDebugEnabled()) {
           if ((now - lastMillis) >= COMPACTION_PROGRESS_LOG_INTERVAL) {
-            LOG.debug("Compaction progress: "
-                + compactionName
-                + " "
-                + progress
-                + String.format(", rate=%.2f kB/sec", (bytesWrittenProgressForLog / 1024.0)
-                    / ((now - lastMillis) / 1000.0)) + ", throughputController is "
-                + throughputController);
+            String rate = String.format("%.2f",
```
Comment (infraio) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java]: Ditto. if (LOG.isDebugEnabled())?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java [hunk 5]
```diff
@@ -330,6 +327,7 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
       throw new InterruptedIOException(
           "Interrupted while control throughput of compacting " + compactionName);
     } finally {
+      ((ShipperListener) writer).beforeShipped();
```
Comment (infraio) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: Add comment here, too?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java (hunk 3 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: Have an existed isDebugEnabled before in line#312 :-)

File: hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java (hunk 5 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java]: OK

File: hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java (hunk 4 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java]: Please see line#449.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 68 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 280 | master passed |
| +1 | compile | 57 | master passed |
| +1 | checkstyle | 82 | master passed |
| +1 | shadedjars | 292 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 241 | master passed |
| +1 | javadoc | 38 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 273 | the patch passed |
| +1 | compile | 64 | the patch passed |
| +1 | javac | 64 | the patch passed |
| +1 | checkstyle | 86 | hbase-server: The patch generated 0 new + 66 unchanged - 1 fixed = 66 total (was 67) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 306 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 868 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 257 | the patch passed |
| +1 | javadoc | 36 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 17098 | hbase-server in the patch failed. |
| +1 | asflicense | 32 | The patch does not generate ASF License warnings. |
| | | 20428 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.regionserver.TestSplitTransactionOnCluster |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.5 Server=18.09.5 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/341 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 1697b6eaaace 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | master / a4738e5184 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/3/testReport/ |
| Max. process+thread count | 4813 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-341/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request merged**